src/search_stream.rs

```diff
@@ -281,7 +297,7 @@ impl<'a, R: io::Read, W: Terminal + Send> Searcher<'a, R, W> {
 
     #[inline(always)]
     fn print_before_context(&mut self, upto: usize) {
-        if self.opts.count || self.opts.before_context == 0 {
+        if self.opts.count || self.opts.files_with_matches || self.opts.before_context == 0 {
```

Comment:
@BurntSushi I think it might not be a bad refactor to add an `opts.just_file = self.opts.count || self.opts.files_with_matches`, which can replace four instances of this or statement; eh?